lib/linked_list.rb

```diff
@@ -19,99 +19,271 @@ def initialize
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
     def add_first(value)
-      raise NotImplementedError
+      @head = Node.new(value, @head)
     end
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
     def search(value)
-      raise NotImplementedError
+      current = @head
+
+      until current.nil?
+        return true if current.data == value
+        current = current.next
+      end
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max
-      raise NotImplementedError
+      current = @head
+      temp = current
+
+      return if current.nil?
+
+      until current.nil?
+        if current.data > temp.data
+          temp = current
+        end
+        current = current.next
+      end
+      return temp.data
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
     def find_min
-      raise NotImplementedError
-    end
+      current = @head
+      temp = current
 
+      return if current.nil?
+
+      until current.nil?
+        if current.data < temp.data
+          temp = current
+        end
+        current = current.next
+      end
+      return temp.data
+    end
+    
+    # delete five (not from assignment)
+    def del_five
+      5.times do |i|
+        if @head.nil?
+          return
+        else
+          @head = @head.next
+        end
+      end
+    end
 
     # method that returns the length of the singly linked list
     def length
-      raise NotImplementedError
+      current = @head
+      count = 0
+
+      return count if current.nil?
+
+      until current.nil?
+        count += 1
+        current = current.next
+      end
+      return count
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
     def get_at_index(index)
-      raise NotImplementedError
+      current = @head
+
+      index.times do
+        return nil if current.nil?
+        current = current.next
+      end
+
+      return current.data
     end
 
     # method to print all the values in the linked list
     def visit
-      raise NotImplementedError
+      current = @head
+      
+      until current.nil?
+        print current.data
+        current = current.next
+      end
     end
 
     # method to delete the first node found with specified value
     def delete(value)
-      raise NotImplementedError
+      current = @head
+      previous = nil
+
+      return nil if current.nil?
+
+      until current.data == value
+        return nil if current.nil?

```

Comment:
The return will never execute, since the `until` would crash the app if `current` is `nil`.